File: apps/backend/app/api/routes/disruptions.py
from fastapi import APIRouter, status
from app.schemas.disruption import DisruptionSimulationRequest, DisruptionSimulationResponse
from app.services.disruption_service import disruption_service

router = APIRouter(prefix="/api", tags=["Disruptions"])


@router.post(
    "/simulations/flight-cancellation",
    response_model=DisruptionSimulationResponse,
    status_code=status.HTTP_201_CREATED,
)
async def simulate_flight_cancellation(request: DisruptionSimulationRequest):
    """
    Trigger a flight cancellation simulation:
    - Validates trip and selected flight segment
    - Marks flight as CANCELLED in database
    - Creates disruption document
    - Emits initial DISRUPTION_DETECTED timeline event
    - Broadcasts event via WebSocket to live clients
    - Dispatches job to autonomous worker service
    """
    result = await disruption_service.simulate_flight_cancellation(
        trip_id=request.trip_id,
        flight_id=request.flight_id,
        reason=request.reason,
    )
    return result


@router.get("/disruptions/{disruption_id}")
def get_disruption(disruption_id: str):
    """
    Retrieve disruption details by disruption ID.
    """
    return disruption_service.get_disruption(disruption_id)

File: apps/frontend/src/api/disruptions.js
import apiClient from './axios'

/**
 * Trigger flight cancellation simulation.
 * Backend updates flight status to CANCELLED, records disruption document,
 * emits DISRUPTION_DETECTED to timeline, broadcasts via WebSocket, and dispatches to worker.
 *
 * @param {Object} params
 * @param {string} params.tripId
 * @param {string} params.flightId
 * @param {string} [params.reason]
 */
export const simulateFlightCancellation = async ({ tripId, flightId, reason }) => {
  return apiClient.post('/api/simulations/flight-cancellation', {
    trip_id: tripId,
    flight_id: flightId,
    reason: reason || 'Flight cancelled by airline due to operational constraints',
  })
}

/**
 * Fetch disruption details by ID.
 */
export const fetchDisruption = async (disruptionId) => {
  if (!disruptionId) throw new Error('Disruption ID is required')
  return apiClient.get(`/api/disruptions/${encodeURIComponent(disruptionId)}`)
}

export default {
  simulateFlightCancellation,
  fetchDisruption,
}
